Addon is crashing with unexpected error, and watchdog is not re-starting it

Add-on version: 2022.5.1
 You are running the latest version of this add-on.
 System: Home Assistant OS 9.5  (amd64 / qemux86-64)
 Home Assistant Core: 2023.1.7
 Home Assistant Supervisor: 2023.01.1


I have noticed this happen a few times over the last month that I've been using the addon (I used to just run it in a self-managed docker container and never had an issue).

I will notice that all of a sudden the addon is stopped, and the logs only contain a cryptic unexpected error.

I have the "watchdog" option turned on, but it doesn't seem to actually revive the addon.  However, hitting "start" manually turns the addon back on without any issues, then it works for a week or so before happening again.


Edit:  I caught the log this time and will paste it below.  It looks like it is happening when the doorbell can't be reached in 6 seconds.  I guess my question is why that kills the addon and why it doesn't get restarted by the watchdog.

```
[cont-init.d] 00-banner.sh: exited 0.
[cont-init.d] 01-log-level.sh: executing... 
[cont-init.d] 01-log-level.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
01/02/2023 22:36:15 [INFO] App Version: 1.0.14
01/02/2023 22:36:15 [INFO] Fetching camera details...
<Front Door:Z1760254EDDD7> Trying again due to error: ReadTimeout(ReadTimeoutError("HTTPConnectionPool(host='192.168.10.53', port=80): Read timed out. (read timeout=6.05)"))
01/02/2023 22:36:26 [INFO] Device type: AD410
01/02/2023 22:36:26 [INFO] Serial number: Z1760254EDDD7
01/02/2023 22:36:26 [INFO] Software version: 1.000.0000000.7.R
01/02/2023 22:36:26 [INFO] Device name: doorbell
01/02/2023 22:38:06 [ERROR] Ping unsuccessful
01/02/2023 22:38:06 [INFO] Exiting app...
[cmd] ./run.sh exited 1
[cont-finish.d] executing container finish scripts...
[cont-finish.d] 99-message.sh: executing... 
-----------------------------------------------------------
                Oops! Something went wrong.

 We are so sorry, but something went terribly wrong when
 starting or running this add-on.
 
 Be sure to check the log above, line by line, for hints.
-----------------------------------------------------------
[cont-finish.d] 99-message.sh: exited 0.
[cont-finish.d] done.
[s6-finish] waiting for services.
[s6-finish] sending all processes the TERM signal.
```

--- a/.github/workflows/publish.yaml
+++ b/.github/workflows/publish.yaml
@@ -3,7 +3,7 @@ name: "Publish"
 on:
   workflow_dispatch:
     branches:
-      - master
+      - main
   release:
     types: [published]
 

--- a/amcrest2mqtt-addon/CHANGELOG.md
+++ b/amcrest2mqtt-addon/CHANGELOG.md
@@ -2,6 +2,7 @@
 
 - Upgrade base image to 13.1.3
 - Switch to [@carefulcomputer fork](https://github.com/carefulcomputer/amcrest2mqtt)
+- Move to service to try to fix #9
 
 **Full Changelog**: https://github.com/ikifar2012/amcrest2mqtt-addon/compare/2022.5.1...2023.3.0
 

--- a/amcrest2mqtt-addon/CHANGELOG.md
+++ b/amcrest2mqtt-addon/CHANGELOG.md
@@ -1,3 +1,10 @@
+# 2023.3.0
+
+- Upgrade base image to 13.1.3
+- Switch to [@carefulcomputer fork](https://github.com/carefulcomputer/amcrest2mqtt)
+
+**Full Changelog**: https://github.com/ikifar2012/amcrest2mqtt-addon/compare/2022.5.1...2023.3.0
+
 # 2022.5.1
 
 - Revert amcrest2mqtt to version 1.0.15 to mitigate issue #4

--- a/amcrest2mqtt-addon/Dockerfile
+++ b/amcrest2mqtt-addon/Dockerfile
@@ -9,7 +9,7 @@ RUN apk add --no-cache \
 python3 \
 py3-pip \
 git
-RUN git clone -b 1.0.15 https://github.com/dchesterton/amcrest2mqtt.git 
+RUN git clone https://github.com/carefulcomputer/amcrest2mqtt.git
 # Copy data for add-on
 COPY run.sh /
 RUN pip install -r /amcrest2mqtt/requirements.txt

--- a/amcrest2mqtt-addon/Dockerfile
+++ b/amcrest2mqtt-addon/Dockerfile
@@ -31,7 +31,7 @@ LABEL \
     org.label-schema.build-date=${BUILD_DATE} \
     org.label-schema.name="amcrest2mqtt" \
     org.label-schema.schema-version="1.0" \
-    org.label-schema.usage="https://github.com/ikifar2012/amcrest2mqtt-addon/blob/master/README.md" \
+    org.label-schema.usage="https://github.com/ikifar2012/amcrest2mqtt-addon/blob/main/README.md" \
     org.label-schema.vcs-ref=${BUILD_REF} \
     org.label-schema.vcs-url="https://github.com/ikifar2012/amcrest2mqtt-addon/" \
     org.label-schema.vendor="Matheson's Home Assistant Addons"

--- a/amcrest2mqtt-addon/Dockerfile
+++ b/amcrest2mqtt-addon/Dockerfile
@@ -6,15 +6,13 @@ ENV LANG C.UTF-8
 SHELL ["/bin/bash", "-o", "pipefail", "-c"]
 # Install Components
 RUN apk add --no-cache \
-python3 \
-py3-pip \
-git
+    python3 \
+    py3-pip \
+    git
 RUN git clone https://github.com/carefulcomputer/amcrest2mqtt.git
 # Copy data for add-on
-COPY run.sh /
 RUN pip install -r /amcrest2mqtt/requirements.txt
-RUN chmod +x run.sh
-CMD [ "./run.sh" ]
+COPY rootfs /
 # ARGS
 ARG BUILD_DATE
 ARG BUILD_REF

--- a/amcrest2mqtt-addon/build.yaml
+++ b/amcrest2mqtt-addon/build.yaml
@@ -1,10 +1,10 @@
 squash: false
 build_from:
-  aarch64: ghcr.io/hassio-addons/base/aarch64:11.1.2
-  amd64: ghcr.io/hassio-addons/base/amd64:11.1.2
-  armhf: ghcr.io/hassio-addons/base/armhf:11.1.2
-  armv7: ghcr.io/hassio-addons/base/armv7:11.1.2
-  i386: ghcr.io/hassio-addons/base/i386:11.1.2
+  aarch64: ghcr.io/hassio-addons/base/aarch64:13.1.3
+  amd64: ghcr.io/hassio-addons/base/amd64:13.1.3
+  armhf: ghcr.io/hassio-addons/base/armhf:13.1.3
+  armv7: ghcr.io/hassio-addons/base/armv7:13.1.3
+  i386: ghcr.io/hassio-addons/base/i386:13.1.3
 codenotary:
   signer: cas@mathesonsteplock.ca
   base_image: codenotary@frenck.dev

--- a/amcrest2mqtt-addon/config.yaml
+++ b/amcrest2mqtt-addon/config.yaml
@@ -1,5 +1,5 @@
 name: amcrest2mqtt
-version: 2022.5.1
+version: 2023.3.0
 slug: amcrest2mqtt-addon
 description: A Home Assistant addon for amcrest2mqtt
 url: https://github.com/ikifar2012/amcrest2mqtt-addon/blob/master/README.md

--- a/amcrest2mqtt-addon/config.yaml
+++ b/amcrest2mqtt-addon/config.yaml
@@ -13,6 +13,7 @@ image: ikifar/amcrest2mqtt-addon-{arch}
 codenotary: cas@mathesonsteplock.ca
 startup: services
 boot: auto
+init: false
 map:
 - ssl:ro
 services:

--- a/amcrest2mqtt-addon/rootfs/etc/services.d/amcrest2mqtt/finish
+++ b/amcrest2mqtt-addon/rootfs/etc/services.d/amcrest2mqtt/finish
@@ -1,0 +1,15 @@
+#!/usr/bin/env bashio
+# ==============================================================================
+# Take down the S6 supervision tree when amcrest2mqtt exits with an error
+# s6-overlay docs: https://github.com/just-containers/s6-overlay
+# ==============================================================================
+
+declare APP_EXIT_CODE=${1}
+
+if [[ "${APP_EXIT_CODE}" -ne 0 ]] && [[ "${APP_EXIT_CODE}" -ne 256 ]]; then
+  bashio::log.warning "Halt add-on with exit code ${APP_EXIT_CODE}"
+  echo "${APP_EXIT_CODE}" > /run/s6-linux-init-container-results/exitcode
+  exec /run/s6/basedir/bin/halt
+fi
+
+bashio::log.info "Service restart after closing"

--- a/amcrest2mqtt-addon/rootfs/etc/services.d/amcrest2mqtt/run
+++ b/amcrest2mqtt-addon/rootfs/etc/services.d/amcrest2mqtt/run
@@ -1,4 +1,10 @@
 #!/usr/bin/env bashio
+# ==============================================================================
+# Start the amcrest2mqtt service
+# s6-overlay docs: https://github.com/just-containers/s6-overlay
+# ==============================================================================
+
+
 
 #set environment variables
 export AMCREST_HOST=$(bashio::config 'amcrest_host')
@@ -20,5 +26,5 @@ else
     bashio::log.warning "MQTT service not available, please ensure the MQTT Broken addon is installed and configured"
 fi
 
-cd /amcrest2mqtt/
-exec /usr/bin/python3 -u src/amcrest2mqtt.py
+cd /amcrest2mqtt/ || exit
+exec /usr/bin/python3 -u src/amcrest2mqtt.py || bashio::exit.nok "Failed to start amcrest2mqtt"

--- a/amcrest2mqtt-addon/rootfs/etc/services.d/amcrest2mqtt/run
+++ b/amcrest2mqtt-addon/rootfs/etc/services.d/amcrest2mqtt/run
@@ -23,7 +23,7 @@ if bashio::services.available "mqtt"; then
     export MQTT_PORT=$(bashio::services "mqtt" "port")
     export MQTT_USERNAME=$(bashio::services "mqtt" "username")
 else
-    bashio::log.warning "MQTT service not available, please ensure the MQTT Broken addon is installed and configured"
+    bashio::log.warning "MQTT service not available, please ensure the MQTT Broker addon is installed and configured"
 fi
 
 cd /amcrest2mqtt/ || exit

